File: Classes/NewGameController.m

// creates and interacts with the dialog in which the user asks for a new game
// some other class must actually display the dialog (as a popover, as it happens)

#import "NewGameController.h"

@interface NewGameController () <UITableViewDataSource, UITableViewDelegate, UIPickerViewDelegate, UIPickerViewDataSource> {
    
}
@property (nonatomic, weak) UITableView* tableView;
@end


@implementation NewGameController

#pragma mark -
#pragma mark View lifecycle

- (void) loadView {
    //UIView* v = [[UIView alloc] initWithFrame:CGRectMake(0,0,320,310+216+100)];
    UIView* v = [[UIView alloc] init];
    UITableView* tv = [[UITableView alloc] initWithFrame:CGRectMake(0,0,320,330) style:UITableViewStyleGrouped];
    [v addSubview:tv];
    tv.dataSource = self;
    tv.delegate = self;
    tv.bounces = NO;
    tv.scrollEnabled = NO;
    self.tableView = tv;
    //UIPickerView* pv = [[UIPickerView alloc] initWithFrame:CGRectMake(0,310,320,216)];
    // new, let the view picker size itself automatically;
    UIPickerView* pv = [[UIPickerView alloc] initWithFrame:CGRectMake(0,0,200,180)]; // suppress annoying error message about wrong height
    [pv sizeToFit];
    //NSLog(@"%f", pv.bounds.size.height);
    CGRect f = pv.frame;
    f.origin.y = tv.frame.size.height;
    pv.frame = f;
    // new, let the view size itself according to its subviews
    v.bounds = CGRectMake(0, 0, 320, tv.frame.size.height + pv.frame.size.height);
    [v addSubview:pv];
    pv.dataSource = self;
    pv.delegate = self;
    pv.showsSelectionIndicator = YES;
    [pv selectRow:[[NSUserDefaults standardUserDefaults] integerForKey:@"Stages"] 
      inComponent:0 animated:NO];
    self.view = v;
    self.contentSizeForViewInPopover = self.view.bounds.size;
    self.modalInPopover = YES;
     // ooooh, this was a leak! fixed
}

- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation {
    // Override to allow orientations other than the default portrait orientation.
    return YES;
}

#pragma mark -
#pragma mark Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    // Return the number of sections.
    return 2;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    NSString* title = nil;
    if (section == 0)
        title = @"Size";
    if (section == 1)
        title = @"Style";
    return title;
}

- (NSString *)tableView:(UITableView *)tableView titleForFooterInSection:(NSInteger)section {
    return nil;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    // Return the number of rows in the section.
    if (section == 0)
        return 3;
    return 2;
}


// Customize the appearance of table view cells.
- (UITableViewCell *)tableView:(UITableView *)tableVieww cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    static NSString *CellIdentifier = @"Cell";
    
    UITableViewCell *cell = [tableVieww dequeueReusableCellWithIdentifier:CellIdentifier];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier];
    }
    
    NSUInteger section = [indexPath section];
    NSUInteger row = [indexPath row];
    
    NSUserDefaults* ud = [NSUserDefaults standardUserDefaults];
    
    if (section == 0) {
        // size
        if (row == 0) {
            cell.textLabel.text = @"Easy";
        } else if (row == 1) {
            cell.textLabel.text = @"Normal";
        } else if (row == 2) {
            cell.textLabel.text = @"Hard";
        }
    } else if (section == 1) {
        if (row == 0) {
            cell.textLabel.text = @"Animals";
        } else if (row == 1) {
            cell.textLabel.text = @"Snacks";
        }
    }
    cell.accessoryType = UITableViewCellAccessoryNone;
    if ([[ud valueForKey:@"Style"] isEqualToString:cell.textLabel.text] ||
        [[ud valueForKey:@"Size"] isEqualToString:cell.textLabel.text])
        cell.accessoryType = UITableViewCellAccessoryCheckmark;
    
    return cell;
}


#pragma mark -
#pragma mark Table view delegate

- (void)tableView:(UITableView *)tv didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
//    NSUInteger section = [indexPath section];
//    NSUInteger row = [indexPath row];
    NSUserDefaults* ud = [NSUserDefaults standardUserDefaults];
    /*
    if (section == 0) {
        // size
        if (row == 0) {
            [ud setValue: @"Easy" forKey: @"Size"];
        } else if (row == 1) {
            [ud setValue: @"Normal" forKey: @"Size"];
        } else if (row == 2) {
            [ud setValue: @"Hard" forKey: @"Size"];
        }
    } else if (section == 1) {
        if (row == 0) {
            [ud setValue: @"Animals" forKey: @"Style"];
        } else if (row == 1) {
            [ud setValue: @"Snacks" forKey: @"Style"];
        }
    }
    */
    NSString* setting = [tv cellForRowAtIndexPath:indexPath].textLabel.text;
    [ud setValue:setting forKey:[self tableView:tv titleForHeaderInSection:indexPath.section]];

    [self.tableView reloadData];
}

#pragma mark -
#pragma mark Picker view support


- (NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
    return [NSString stringWithFormat: @"%i Stage%@", row+1, (row > 0 ? @"s" : @"")];
}

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    return 9;
}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component {
    [[NSUserDefaults standardUserDefaults] setObject:@(row) forKey:@"Stages"];
}

- (CGFloat)pickerView:(UIPickerView *)pickerView rowHeightForComponent:(NSInteger)component {
    return 25.0;
}

@end

File: Classes/NewGameController.swift


import UIKit
import Swift

private let cellid = "Cell"
private let headerid = "Header"

class NewGameController : UIViewController {
    weak var tableView : UITableView?
    
    init () {
        super.init(nibName: nil, bundle: nil)
        self.edgesForExtendedLayout = []
    }
    
    required init(coder: NSCoder) {
        fatalError("NSCoding not supported")
    }
    
    override func viewDidLoad() {
        
        super.viewDidLoad()

        let v = self.view
        v?.backgroundColor = UIColor.white
        
        // initial table height just an estimate, real height will be determined later
        let tableHeight : CGFloat = (onPhone ? 120 : 300)
        let tv = UITableView(frame:CGRect(x: 0,y: 0,width: 320,height: tableHeight), style:.plain)
        
        v?.addSubview(tv)
        tv.dataSource = self
        tv.delegate = self
        tv.bounces = false
        tv.isScrollEnabled = false
        tv.register(UITableViewCell.self, forCellReuseIdentifier: cellid)
        tv.register(UITableViewHeaderFooterView.self, forHeaderFooterViewReuseIdentifier: headerid)
        self.tableView = tv
        tv.translatesAutoresizingMaskIntoConstraints = false
        
        let pv = UIPickerView()
        pv.translatesAutoresizingMaskIntoConstraints = false
        pv.dataSource = self
        pv.delegate = self
        v?.addSubview(pv)
        // no table height constraint yet
        NSLayoutConstraint.activate([
            NSLayoutConstraint.constraints(withVisualFormat: "H:|[tv]|", options: [], metrics: nil, views: ["tv":tv]),
            NSLayoutConstraint.constraints(withVisualFormat: "H:|[pv]|", options: [], metrics: nil, views: ["pv":pv]),
            NSLayoutConstraint.constraints(withVisualFormat: "V:[tv]-(0)-[pv]", options: [],
                metrics: nil,
                views: ["tv":tv, "pv":pv])
            ].flatMap{$0})
        pv.showsSelectionIndicator = true
        pv.selectRow(ud.integer(forKey: Default.lastStage), inComponent: 0, animated: false)

        self.preferredContentSize = CGSize(width: 320, height: tv.frame.size.height + pv.frame.size.height)
    }
    
    // determine actual table height constraint
    var didUpdateConstraints = false
    override func updateViewConstraints() {
        if !didUpdateConstraints {
            didUpdateConstraints = true
            var h : CGFloat = 0
            let tv = self.tableView!
            let secs = tv.numberOfSections
            for sec in 0..<secs {
                h += tv.rectForHeader(inSection: sec).height
                let rows = tv.numberOfRows(inSection: sec)
                for row in 0..<rows {
                    h += tv.rectForRow(at: IndexPath(row: row, section: sec)).height
                }
            }
            NSLayoutConstraint.activate([
                NSLayoutConstraint.constraints(withVisualFormat: "V:|-(0)-[tv(tableHeight)]", options: [],
                                               metrics: ["tableHeight":h],
                                               views: ["tv":tv])
                ].flatMap{$0})
        }
        super.updateViewConstraints()
    }
}

extension NewGameController : UITableViewDelegate, UITableViewDataSource {
    func numberOfSections(in tableView: UITableView) -> Int {
        return onPhone ? 1 : 2 // on iPhone, omit second (Size) section: there is just one size
    }
    
    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        let v = tableView.dequeueReusableHeaderFooterView(withIdentifier: headerid)!
        if v.viewWithTag(99) == nil {
            let lab = UILabel()
            lab.tag = 99
            lab.font = UIFont.systemFont(ofSize: 20, weight: .bold)
            lab.translatesAutoresizingMaskIntoConstraints = false
            v.contentView.addSubview(lab)
            NSLayoutConstraint.activate([
                lab.topAnchor.constraint(equalTo: v.contentView.topAnchor),
                lab.bottomAnchor.constraint(equalTo: v.contentView.bottomAnchor),
                lab.leadingAnchor.constraint(equalTo: v.contentView.layoutMarginsGuide.leadingAnchor)
            ])
        }
        let lab = v.viewWithTag(99) as! UILabel
        lab.text = section == 0 ? Default.style : Default.size
        return v
    }
    
    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 30
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        switch section {
        case 0:
            return 2
        case 1:
            return 3
        default:
            return 0
        }
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: cellid, for:indexPath)
        
        let section = indexPath.section
        let row = indexPath.row
        
        switch section {
        case 0:
            cell.textLabel!.text = Styles.styles()[row]
        case 1:
            cell.textLabel!.text = Sizes.sizes()[row]
        default:
            cell.textLabel!.text = "" // throwaway
        }
        cell.textLabel?.font = UIFont.systemFont(ofSize: 17, weight: .regular)
        
        cell.accessoryType = .none
        let currentDefaults = [ud.string(forKey: Default.style), ud.string(forKey: Default.size)]
        if currentDefaults.contains(where: {$0 == cell.textLabel!.text}) {
            cell.accessoryType = .checkmark
        }

        return cell
    }

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if let setting = tableView.cellForRow(at: indexPath)?.textLabel?.text {
            ud.set(setting, forKey: indexPath.section == 0 ? Default.style : Default.size)
            tableView.reloadData()
        }
    }
}

extension NewGameController : UIPickerViewDelegate, UIPickerViewDataSource {
    
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 1
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return 9
    }
    
    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        return "\(row+1) Stage" + ( row > 0 ? "s" : "")
    }
    
    func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
        ud.set(row, forKey:Default.lastStage)
    }
    
    func pickerView(_ pickerView: UIPickerView, rowHeightForComponent component: Int) -> CGFloat {
        return 35
    }



}
